src/cfg/File.cc

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "cfg/File.h"
+#include "parser/Tokenizer.h"
+#include "sbuf/Stream.h"
+
+#if HAVE_SYS_STAT_H
+#include <sys/stat.h>
+#endif
+
+Cfg::File::~File()
+{
+    if (fd) {
+        if (isPipe)
+            pclose(fd);
+        else
+            fclose(fd);
+    }
+}
+
+void
+Cfg::File::tryLoadFile()
+{
+    if (filePath[0] == '!' || filePath[0] == '|') {
+        isPipe = true;
+        filePath.erase(0,1);
+    }
+
+    struct stat data;
+    memset(&data, 0, sizeof(data));
+
+    if (::stat(filePath.c_str(), &data) != 0) {
+        int xerrno = errno;
+        throw TextException(ToSBuf("configuration file ", filePath, " error: ", xstrerr(xerrno)), lineInfo());
+    }
+
+    if (S_ISFIFO(data.st_mode))
+        isPipe = true;
+
+    debugs(3, 2, "Loading " << (isPipe ? "pipe" : "file") << " " << filePath);
+
+    if (data.st_size == 0)
+        return; // optimization: dont bother reading empty files
+
+    if (isPipe && !(fd = popen(filePath.c_str(), "r")))
+        throw TextException(ToSBuf("configuration pipe :", filePath, " not found"), lineInfo());
+
+    else if (!(fd = fopen(filePath.c_str(), "r")))
+        throw TextException(ToSBuf("configuration file :", filePath, " not found"), lineInfo());
+
+#if _SQUID_WINDOWS_
+    setmode(fileno(fd), O_TEXT);
+#endif
+
+    // try to load the entire file into parseBuffer
+    off_t fileSz = data.st_size;
+    off_t readSz = 0;
+    while (fileSz - readSz) {
+        SBuf parseBuffer;
+
+        auto len = fileSz - readSz;
+        // limit at 1/2 max capacity, so we can combine two SBuf later
+        if (len > SBuf::maxSize>>1)
+            len = SBuf::maxSize>>1;
+
+        auto *p = parseBuffer.rawAppendStart(len);
```

Comment:
Dropping this limit results in SBuf exceptions loading perfectly valid ACL files. There are multiple configuration parsers, several of which do not perform exception handling cleanly. This will result in large sections of configuration being dropped or rejected as "Bungled".

SBuf::maxSize is a bitmask of bits that can be used before the size value hits overflow bugs in SBuf code. It also serves dual-purpose as a size value itself. If you insist on the performance regression I can replace these with slow division operations.

Reply:
This all got dropped in 8d74339 when st_stat usage disappears. For now we always allocate half-sized SBuf as read buffer for this loop. Memory allocation optimizations are left for future work.